Inventory.md

```diff
@@ -0,0 +1,173 @@
+
+We can trivially re-encode addresses to work with the wallet core and re-decode them later
+(when checking if they are ours or not).
+
+We might be able to treat addresses more abstractly and avoid this.
+
+### Active slot coeff
+
+While missing in Byron, it will exist in Shelley. We could set it to 100% internally in the wallet. Should be a problem at all. ✅
+
+### Superfluous prevBlockHash in BlockHeader
+
+With a good abstraction, or dropped Jörmungandr support we could remove the prevBlockHash field in BlockHeader, but it does no damage there. ✅
+
+### Point ByronBlock vs BlockHeader
+
+We can probably get by with magic values to represent the `Origin`-case. ✅
+
+A way to treat points (like ourobouros-network's`Point ByronBlock`) more abstract in the wallet could potentially be needed to avoid sabotaging other abstractions.
+
+### SlotId vs SlotNo
+
+When slotsPerEpoch cannot change the conversion between SlotId and SlotNo is simple.
+
+But to support protocol parameter updates, we should consider making the core wallet logic deal with SlotNo for the Haskell node.
+
+For Byron, we can assume they are static. ✅
+
+### Epoch Bounday Blocks
+
+I imagine having two blocks with the same slot might break assumptions in our DB-rollbacks, but I don't know.
+
+A simple solution would be to filter them out. ✅
+
+For mainnet Shelley, EBBs do provide an easy way of knowing when a new epoch starts without having to keep track of protocol parameters (and when slots are counted from genesis). This would be an argument for not filtering them out. We will however most likely be maintaining up-to-date protocol parameters anyway, and would be able to tell when we're entering a new epoch through the protocol-params and the current slotNo.
+
+### No rollbacks further than "k"
+
+With the Haskell node we will be able to say that old enough state/checkpoints are immutable and can never be rolled back. We might want to change which checkpoints we store in DB. There is no point in keeping multiple checkpoints in the immutable part of the chain.
+
+### Delegation features
+
+We cannot implement delegation features. We could
+1. Keep the API the same, and use dummy implementations internally. E.g return an empty list of stake pools. E.g say that wallets have 0 account balance. I think this should work without any hustle.
+2. Make the wallet modular enough to not have to do the above. I am not sure how difficult this would be.
+
+## Other, fuzzier difficulties of integration
+
+I tried to get a wallet executable working with the Haskell node and discovered new kinds of difficulties.
+
+- Chain parameters are not in the Haskell genesis block. The wallet will need to hard-code or read them from a config file at startup.
+- The genesis block does not have to be treated differently because of above. The jörmungandr wallet currently treats it completely separate.
+- NetworkLayer and `follow` seem like the wrong abstractions to work with both jörmungandre and Haskell.
```

Comment:
why `follow` is wrong? could you give more details?

Reply:
Saying that something is "the wrong abstractions" is one thing. Explaining *why* is actually much more relevant. 
To me, it seems like a relatively good abstraction to work with both backends (modulo the bits related to stake-pools that were added a bit in a hurry and should probably not be part of the network layer in the first place...)

Reply:
I was in a hurry too when making the final prototype and writing this section — sorry!

I couldn't figure out how to implement `networkTip` and `nextBlocks` in a good way with the Haskell node. I will try to provide some reasoning in hindsight:

The `NetworkLayer` (and therefore cardano-wallet-jörmungandr) implements
```haskell
    { nextBlocks
        :: Cursor target
        -> ExceptT ErrGetBlock m (NextBlocksResult target block)
```
which the `follow` function calls repeatedly with different `Cursor target`.

The Haskell chain sync protocol is stateful in the sense that the node knows where we are on the chain and will tell us when to roll forward and when to roll back automatically. A  `nextBlocks` without the `Cursor target` argument (local tip) would work better. But I thought the most straight-forward abstraction for the Haskell node would be:

```haskell
follow 
    :: [point] -- points/BlockHeaders we know about
    -> (RollBackOrForward block -> IO ()) 
    -> IO () -- syncs forever
```
which we could put inside `NetworkLayer. `[src](https://github.com/input-output-hk/cardano-wallet/blob/ed10bfeabac2e184a3d090d875846091249a3288/lib/core/src/Cardano/Wallet/Network.hs#L65-L68), and note this is different from our current `follow` in `Cardano.Wallet.Network`.

It was my intention this `follow` would also work with `cardano-wallet-jormungandr`. But it does not! [`monitorStakePools`](https://github.com/input-output-hk/cardano-wallet/blob/cc7b767608399279fe1f88f5be2ed84268af3d35/lib/core/src/Cardano/Pool/Metrics.hs#L155-L229) relies on being able to tell the current `Cardano.Wallet.Network.follow` to `RetryImmediately`. This is to avoid *jörmungandr-specific* race-conditions when fetching the current stake-distribution. We can't tell the Haskell node to give us a block it has already sent us. Or even if we can with some trickery, it would just be unnecessary. Which is why I didn't  think the current `monitorStakePools` or `Cardano.Wallet.Network.follow` would be good abstractions, nor `nextBlocks`.

NOTE: I believe the chain-sync protocol might be meant to be used with some `TVar` thing. That might invalidate my reasoning *slightly*. I believe @KtorZ should be able tell easily though.

Reply:
I replaced the section with a slightly modified version of this comment, and updated the summary-section.